authenticators/jwt_from_extra_provider.go

```diff
@@ -0,0 +1,85 @@
+package authenticators
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/arrikto/oidc-authservice/common"
+	goidc "github.com/coreos/go-oidc"
+)
+
+const bearerPrefix = "Bearer "
+
+type JWTFromExtraProviderAuthenticator struct {
+	headerName   string
+	issuer       string
+	isserName    string
+	remoteKeySet goidc.KeySet
+}
+
+// This is not a full implementation of OIDC
+// It *just* checks the token against the keys of an extra provider.
+func NewJWTFromExtraProviderAuthenticator(
+	headerName, issuer, issuerName, providerURL string) Authenticator {
+	return &JWTFromExtraProviderAuthenticator{
+		headerName:   headerName,
+		issuer:       issuer,
+		isserName:    issuerName,
+		remoteKeySet: goidc.NewRemoteKeySet(context.Background(), providerURL+"/keys"),
+	}
+}
+
+func (s *JWTFromExtraProviderAuthenticator) Authenticate(w http.ResponseWriter, r *http.Request) (*common.User, bool, error) {
+	logger := common.RequestLogger(r, "JWT extra token authenticator")
+
+	if len(r.Header[s.headerName]) != 1 {
+		return nil, false, nil
+	}
+
+	rawIDToken := strings.TrimPrefix(r.Header[s.headerName][0], bearerPrefix)
+
+	// Validate token
+	verifier := goidc.NewVerifier(s.issuer, s.remoteKeySet, &goidc.Config{SkipClientIDCheck: true})
```

Comment:
This is technically breaking spec (https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication)

> "The Authorization Server MUST verify that it is an intended audience for the token. The Audience SHOULD be the URL of the Authorization Server's Token Endpoint."

So i guess technically we should include in the JWT an aud with a list of all cluster endpoints from all the different authservices, and on each validate that this is present in the JWT

But I can't really think of any downside to skipping this check - because we are validating the signature of the token anyway (by sharing auth) - so what function does the audience parameter even have?

Reply:
As of now, the issuer has the url of the Authorization Server, so we do this check.
We don't check the clientID, which is "oidc-auth" for us. I didn't see it as important, as it can be easily faked, it's not something that can really validate that the JWT is legit, only the check against the public key can, so that's why I disabled it.
I can add this as an env var and check it anyways, let me change this.